FWIO/RNTuple/plugins/RNTupleOutputFile.cc

```diff
@@ -0,0 +1,583 @@
+#include "FWCore/ParameterSet/interface/Registry.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/GlobalIdentifier.h"
+#include "FWCore/Utilities/interface/ConvertException.h"
+
+#include "DataFormats/Provenance/interface/ParentageRegistry.h"
+#include "DataFormats/Provenance/interface/ProductRegistry.h"
+#include "DataFormats/Provenance/interface/FileID.h"
+#include "DataFormats/Provenance/interface/ProcessHistoryRegistry.h"
+#include "DataFormats/Provenance/interface/Provenance.h"
+#include "DataFormats/Provenance/interface/IndexIntoFile.h"
+#include "DataFormats/Provenance/interface/ProductDependencies.h"
+
+#include "IOPool/Common/interface/getWrapperBasePtr.h"
+
+#include "TFile.h"
+#include "ROOT/RNTuple.hxx"
+#include "ROOT/RNTupleWriter.hxx"
+#include <string>
+#include <optional>
+#include <map>
+
+namespace {
+  ROOT::RCompressionSetting::EAlgorithm::EValues convert(edm::rntuple::CompressionAlgos iAlgos) {
+    using namespace edm::rntuple;
+    using namespace ROOT;
+    switch (iAlgos) {
+      case CompressionAlgos::kLZMA:
+        return RCompressionSetting::EAlgorithm::kLZMA;
+      case CompressionAlgos::kZSTD:
+        return RCompressionSetting::EAlgorithm::kZSTD;
+      case CompressionAlgos::kZLIB:
+        return RCompressionSetting::EAlgorithm::kZLIB;
+      case CompressionAlgos::kLZ4:
+        return RCompressionSetting::EAlgorithm::kLZ4;
+    }
+    return RCompressionSetting::EAlgorithm::kZSTD;
+  }
+}  // namespace
+
+using namespace ROOT::Experimental;
+namespace edm {
+  RNTupleOutputFile::RNTupleOutputFile(std::string const& iFileName,
+                                       FileBlock const& iFileBlock,
+                                       SelectedProductsForBranchType const& iSelected,
+                                       Config const& iConfig,
+                                       bool anyProductProduced)
+      : file_(iFileName.c_str(), "recreate", ""),
+        wrapperBaseTClass_(TClass::GetClass("edm::WrapperBase")),
+        selectorConfig_(iConfig.selectorConfig),
+        dropMetaData_(iConfig.dropMetaData) {
+    setupRuns(iSelected[InRun], iConfig);
+    setupLumis(iSelected[InLumi], iConfig);
+    setupEvents(iSelected[InEvent], iConfig, anyProductProduced);
+    setupPSets(iConfig);
+    setupParentage(iConfig);
+    setupMetaData(iConfig);
+
+    auto const& branchToChildMap = iFileBlock.productDependencies().childLookup();
+    for (auto const& parentToChildren : branchToChildMap) {
+      for (auto const& child : parentToChildren.second) {
+        productDependencies_.insertChild(parentToChildren.first, child);
+      }
+    }
+  }
+
+  namespace {
+    std::string fixBranchName(std::string const& iName) {
+      //need to remove the '.' at the end of the branch name
+      return iName.substr(0, iName.size() - 1);
+    }
+
+    void noSplitField(ROOT::Experimental::RFieldBase& iField) {
```

Comment:
This function would benefit from some explanation in comments.

Reply:
A long explanation was added.